core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlUtils.java

```diff
@@ -20,62 +20,46 @@
 import org.apache.calcite.plan.InnerJoinTraitDef;
 import org.apache.calcite.plan.RelTrait;
 import org.apache.calcite.plan.RelTraitSet;
-import org.apache.calcite.rel.RelNode;
 import org.apache.calcite.rel.core.Filter;
 import org.apache.calcite.rel.core.Project;
-import org.apache.calcite.rel.logical.LogicalJoin;
 import org.apache.calcite.rex.RexCall;
 import org.apache.calcite.rex.RexInputRef;
 import org.apache.calcite.rex.RexNode;
 import org.apache.calcite.rex.RexOver;
 import org.apache.calcite.rex.RexVisitorImpl;
-import org.apache.calcite.util.Util;
+
+import com.google.common.collect.ImmutableSet;
+
+import java.util.HashSet;
+import java.util.Set;
 
 /**
  * Utility class for rel2sql package.
  */
 public class RelToSqlUtils {
 
-  /**
-   * Returns whether an operand is Analytical Function by traversing till next project rel
-   * For ex, FilterRel e1 -> FilterRel e2 -> ProjectRel p -> TableScan ts
-   * Here, we are traversing till ProjectRel p to check whether an operand of FilterRel e1
-   * is Analytical function or not.
-   */
-  private static boolean isOperandAnalyticalInFollowingProject(RelNode rel,
-      Integer rexOperandIndex) {
-    if (rel instanceof Project) {
-      return (((Project) rel).getProjects().size() - 1) >= rexOperandIndex
-          && isAnalyticalRex(((Project) rel).getProjects().get(rexOperandIndex));
-    } else if (rel.getInputs().size() > 0) {
-      return isOperandAnalyticalInFollowingProject(rel.getInput(0), rexOperandIndex);
-    }
-    return false;
+  private RelToSqlUtils() {
   }
 
-  /** Returns whether an Analytical Function is present in filter condition. */
-  protected boolean hasAnalyticalFunctionInFilter(Filter rel, RelNode input) {
-    AnalyticalFunctionFinder finder = new AnalyticalFunctionFinder(true, input);
-    try {
-      rel.getCondition().accept(finder);
-      return false;
-    } catch (Util.FoundOne e) {
+  /**
+   * Checks whether the filter can be a QUALIFY clause.
+   * @param filter can be WHERE/HAVING/QUALIFY
+   * @return true when filter contains a window function or an alias of an window function
+   */
+  public static boolean isQualifyFilter(Filter filter) {
+    if (filter.containsOver()) {
       return true;
     }
-  }
-
-  /** Returns whether an Analytical Function is present in joins.*/
-  protected boolean hasAnalyticalFunctionInJoin(RelNode input) {
-    if (input instanceof LogicalJoin && input.getInput(0) instanceof Project) {
-      return isAnalyticalFunctionPresentInProjection((Project) input.getInput(0));
+    if (!Project.class.isAssignableFrom(filter.getInput().getClass())) {
+      return false;
```

Comment:
Is it the same?
```
if (!(filter.getInput() instanceof Project)) {
  return false;
}
```
if so, i think this is more readable

Reply:
changed